### PR TITLE
Pspin tensorflow potential

### PR DESCRIPTION
**Pull request by smcantab**
this PR is really just a thread to discuss and test the implementation of the pspin spherical potential using TensorFlow (TF) as I plan to move these implementations into a separate repository more dedicated to ML. 

I have produced an implementation of the potential that produces the same results as the existing pele (PL) implementation and I have done some very minimal/preliminary profiling on getEnergyGradient calls:

On 4 cpu threads for TF (although 'top' only reports the use of 2 cores, but I am using a vm on my mac so not sure what the vm is doing) 

* the overhead is substantial for small system sizes. 
* For N=100 the two methods take the same time. 
* For N=150 TF is almost 3x faster. 
* For N=200 TF is more than 3x faster and so on.

So I expect that on more threads and/or GPUs the TF implementation to be a lot faster.

I have made certain choices in the implementation so it is possible that things can be tweaked to gain even more speed, any suggestion is more than welcome.

Next I am going to try this within the pele minimizers and then I will proceed to moving the potential to a new repo under the pele organization.

### CODE REVIEW

**Comment by smcantab** [comment]
PS only look at playground/pspin/pspin_potential_tf.py, the rest is from a separate PR that needs to be merged

**Comment by smcantab** [comment]
now it's implemented for arbitrary p

**Comment by smcantab** [comment]
I tested the potential on GPUs, but the overhead due to the repeated variable assigments are about 90% of the computation time for N=200, while the computation itself is lightning fast. So I have posted this question on task overflow, let's hope for some good expert advise

https://stackoverflow.com/questions/34747982/tensorflow-share-value-for-two-different-variables-within-same-operation

**Comment by smcantab** [comment]
I got a nice suggestion from stackoverflow. Considering all changes from the first implementation we are doing much better now: the TF implementation now runs as fast/faster than the cpu implementation for N=50-60 already when minimizing with pele::lbfgs, which is really good I think considering all the overheads due to the repeated initialization of tensorflow and python wrapping of the potential

**Comment by kjs73** [comment]
That sounds very good! Would we save much by using their optimizer, or can we get the job done with e.g. pele::lbfgs?

**Comment by smcantab** [comment]
I have started playing (see last commit) with ML models to see how they work. Turns out that the tutorials on tensorflow's documentation are not very good, instead these tutorials, adapted from theano, https://github.com/nlintz/TensorFlow-Tutorials seem to give better working models. 

For instance the TF uses an explicitely typed cross_entropy definition, while it turns out that if you don't use tf.nn.softmax_cross_entropy_with_logits you get all sorts of junk. 

TF seems to like NaN quite a lot.

The MNIST SoftMax regression model seems to work well, except that if the batch size is >3000 images (the whole of MNIST is 55k) the loss function returns NaN. Other than that LBFGS and FIRE converge nicely.

I have also (stupidly) gone straight into implementing a convNN of MNIST but that seems to be very unstable and minimizations quickly returns NaN, even with TF native minimizers if one does not do SGD. It almost seems that one can minimize only one step: if you don't change the batch after each step, then minimization becomes really hard with the gradient doing all sorts of crazy things. 

Maybe the other examples that I pointed to  work better for a convNN but I think that for now will step back try a simpler 3-layer NN or maybe the deep (2 hidden layers) feed-forward version.

It is also unclear to me weather using dropout is compatible with our application. Any thoughts/suggestions are welcome.

**Comment by smcantab** [comment]
although most of this has moved to lernscapes there are some bug fixes for pele, so I'll wait for  PR #140 to be merged so that I can rebase this one, get rid of all the tensorflow related files and merge only the remaining bug fixes. Tests fail because of a change in the pspin potential definition, I need to decide whether to change the definition or the tests :)
